trio/_unix_pipes.py

```diff
@@ -1,120 +1,158 @@
-                    try:
-                        total_sent += os.write(self._pipe, remaining)
-                    except BrokenPipeError as e:
-                        await _core.checkpoint()
-                        raise _core.BrokenResourceError from e
-                    except BlockingIOError:
-                        await self.wait_send_all_might_not_block()
+    async def aclose(self):
+        await self._fd_holder.aclose()
 
-    async def wait_send_all_might_not_block(self) -> None:
-        if self._closed:
-            await _core.checkpoint()
-            raise _core.ClosedResourceError("This pipe is already closed")
-
-        try:
-            await _core.wait_writable(self._pipe)
-        except BrokenPipeError as e:
-            # kqueue: raises EPIPE on wait_writable instead
-            # of sending, which is annoying
-            # also doesn't checkpoint so we have to do that
-            # ourselves here too
-            await _core.checkpoint()
-            raise _core.BrokenResourceError from e
-
-
-class PipeReceiveStream(_PipeMixin, ReceiveStream):
-    """Represents a receive stream over an os.pipe object."""
+    def fileno(self):
+        return self._fd_holder.fd
 
-    async def receive_some(self, max_bytes: int) -> bytes:
-        if self._closed:
-            await _core.checkpoint()
-            raise _core.ClosedResourceError("this pipe is already closed")
 
-        if not isinstance(max_bytes, int):
-            await _core.checkpoint()
-            raise TypeError("max_bytes must be integer >= 1")
+class PipeReceiveStream(ReceiveStream):
+    """Represents a receive stream over an os.pipe object."""
 
-        if max_bytes < 1:
-            await _core.checkpoint()
-            raise ValueError("max_bytes must be integer >= 1")
+    def __init__(self, fd: int):
+        self._fd_holder = _FdHolder(fd)
+        self._conflict_detector = ConflictDetector(
+            "another task is using this pipe"
+        )
 
-        while True:
-            try:
-                await _core.checkpoint_if_cancelled()
-                data = os.read(self._pipe, max_bytes)
-            except BlockingIOError:
-                await _core.wait_readable(self._pipe)
-            else:
-                await _core.cancel_shielded_checkpoint()
-                break
+    async def receive_some(self, max_bytes: int) -> bytes:
+        async with self._conflict_detector:
+            if not isinstance(max_bytes, int):
+                raise TypeError("max_bytes must be integer >= 1")
+
+            if max_bytes < 1:
+                raise ValueError("max_bytes must be integer >= 1")
+
+            while True:
+                try:
+                    data = os.read(self._fd_holder.fd, max_bytes)
+                except BlockingIOError:
+                    await _core.wait_readable(self._fd_holder.fd)
+                except OSError as e:
+                    await _core.cancel_shielded_checkpoint()
```

Comment:
Is this still meant to be here? You already did a full checkpoint when acquiring the conflict detector

Reply:
Good catch, thanks :-)